posts/promql-guide.md

```diff
@@ -0,0 +1,159 @@
+
+`!~` - does not regex match
+
+![](./images/promql-guide/substring.png)
+
+If you're interested in learning more, [here](https://docs.python.org/3/library/re.html) are the docs on Regex.
+
+### Range Vectors
+
+_Contain data going back in time._
+
+Recall: Only Instant Vectors can be graphed. You'll soon be able to see how to visualize Range Vectors using functions.
+
+`http_requests_total[5m]`
+
+You can also use (s, m, h, d, w, y) to represent (seconds, minutes, hours, ...) respectively.
+
+## Important functions
+
+### For Range Vectors
+
+_You'll notice that we're able to graph all these functions. Since only Instant Vectors can be graphed, they take a Range Vector as a parameter and return a Instant Vector._
+
+`rate(http_requests_total[5m])` - increase of `http_requests_total` averaged over the last 5 minutes. Use this when alerting because it creates a smooth graph.
+
+![](./images/promql-guide/rate.png)
+
+`irate(http_requests_total[5m])` - looks at the 2 most recent samples (up to 5 minutes in the past), rather than averaging like `rate`. You'll notice this creates a sharp graph, so you shouldn't alert on this. _Alert fatigue is real._
+
+![](./images/promql-guide/irate.png)
+
+`increase(http_requests_total[1h])` - # of http requests in the last hour. This is equal to the `rate` * # of seconds.
+
+![](./images/promql-guide/increase.png)
+
+These are a small fraction of the functions, just what we found most popular. You can find the rest [here](https://prometheus.io/docs/prometheus/latest/querying/functions/).
+
+### For Instant Vectors
+
+You'll notice that `rate(http_requests_total[5m])` above provides a ton of data. You can filter that data using your labels, but you can also look at your system as a whole using `sum` (or do both).
+
+`sum(rate(http_requests_total[5m]))`
+
+You can also use `min`, `max`, `avg`, `count`, and `quintile` similarly.
+
+![](./images/promql-guide/sum-rate.png)
+
+This tells you how many total HTTP requests there are, but some are obviously more indicative than others.
+
+`sum by(status_code) (rate(http_requests_total[5m]))`
+
+You can also use `without` rather than `by` to sum on everything not passed as a parameter to without.
+
+![](./images/promql-guide/sum-by-rate.png)
+
+Now, you can see the difference between each status code.
+
+### Offset
+
+You can use `offset` to change the time for Instant and Range Vectors. This can be helpful when comparing current to past usage when determining to trigger an alert.
+
+`sum(rate(http_requests_total[5m] offset 5m))`
+
+Remember to put `offset` directly after the selector.
+
+## Operators
+
+Operators can be used between scalars, vectors, or a mix of the two. Operations between vectors expect to be able to find matching elements for each element on the other side (also known as one-to-one matching), unless otherwise specified.
+
+There are Arithmetic (+, -, \*, /, %, ^), Comparison (==, !=, >, <, >=, <=) and Logical (and, or, unless) operators.
```

Comment:
Not sure if it's worth listing out what each operator does here.

Reply:
Are you saying it would be better to explain what each individual operator does or to leave this section out entirely?

I assume if they're trying to learn Prometheus, they've gotten past the point of learning this.

Reply:
I was thinking more like a list:
`+` - Addition
`-` - 

But yeah that takes up a ton of space. Maybe just a link to the operators page in the docs.